language/stdlib/modules/gas_schedule.mvir

```diff
@@ -0,0 +1,324 @@
+          Cost { cpu: 917, storage: 1 }
+        );
+        Vector.push_back<Self.Cost>(
+          &mut instruction_table,
+          Cost { cpu: 774, storage: 1 }
+        );
+        Vector.push_back<Self.Cost>(
+          &mut instruction_table,
+          Cost { cpu: 1119, storage: 1 }
+        );
+        Vector.push_back<Self.Cost>(
+          &mut instruction_table,
+          Cost { cpu: 29, storage: 1 }
+        );
+        Vector.push_back<Self.Cost>(
+          &mut instruction_table,
+          Cost { cpu: 41, storage: 1 }
+        );
+        Vector.push_back<Self.Cost>(
+          &mut instruction_table,
+          Cost { cpu: 10, storage: 1 }
+        );
+
+        move_to_sender<T>(T {
+            instruction_schedule: move(instruction_table),
+            native_schedule: Vector.empty<Self.Cost>(),
+        });
+        return;
+    }
+
+    public new_cost(cpu: u64, storage: u64): Self.Cost {
+      return Cost {cpu: move(cpu), storage: move(storage) };
+    }
+
+    // Add a cost to the cost table. Whether the cost is native or not is specified by the
+    // `is_instruction_cost` variable that is passed in.
+    // Note: This adds the instruction onto the end of the vector.
+    public add_cost(is_instruction_cost: bool, cost: Self.Cost) acquires T {
+        let gas_schedule: &mut Self.T;
+        assert(get_txn_sender() == 0xA550C18, 1);
+
+        gas_schedule = borrow_global_mut<T>(0xA550C18);
+        if (move(is_instruction_cost)) {
+            Vector.push_back<Self.Cost>(&mut move(gas_schedule).instruction_schedule, move(cost));
+        } else {
+            Vector.push_back<Self.Cost>(&mut move(gas_schedule).native_schedule, move(cost));
+        }
+        return;
+    }
+
+    // Update the instruction at `index` with new cost `cost`. The table in which this cost should
+    // be updated is specified by `is_instruction_cost`.
+    public update_schedule(is_instruction_cost: bool, index: u64, cost: Self.Cost) acquires T {
+        let gas_schedule: &mut Self.T;
+        let gas_table: &mut Vector.T<Self.Cost>;
+        let new_cost: &mut Self.Cost;
+        let len: u64;
+        assert(get_txn_sender() == 0xA550C18, 2);
+
+        gas_schedule = borrow_global_mut<T>(0xA550C18);
+
+        // Grab the appropriate table
+        if (move(is_instruction_cost)) {
+            gas_table = &mut copy(gas_schedule).instruction_schedule;
+        } else {
+            gas_table = &mut copy(gas_schedule).native_schedule;
+        }
+
+        // Make sure that we aren't over indexing
+        len = Vector.length<Self.Cost>(freeze(copy(gas_table)));
+        assert(copy(index) < move(len), 3);
+
+        // Update the cost at `index`
+        new_cost = Vector.borrow_mut<Self.Cost>(move(gas_table), move(index));
+        *copy(new_cost) = move(cost);
+
+        return;
+    }
+
+    public has_gas_schedule_resource(): bool {
+        return exists<T>(0xA550C18);
+    }
+
+    public table_sizes(): u64 * u64 acquires T {
```

Comment:
Nit: perhaps separate procedures for each table size would make sense? That way, no one has to worry about the order of the tuple.

Reply:
I went back-and-forth on this, and wasn't sure which one to go with.

I'll split into two :)